### PR TITLE
backport: Merge bitcoin#20816,20844,20584,14501,18669

### PR DESCRIPTION
**Pull request by vijaydasmp**

### CODE REVIEW

**Comment by vijaydasmp** [comment]
waiting for https://github.com/dashpay/dash/pull/5490 to merge

File: src/net.cpp [hunk 1]
```diff
@@ -1404,6 +1404,7 @@ void CConnman::CalculateNumConnectionsChangedStats()
         LOCK(pnode->cs_vRecv);
         for (const mapMsgCmdSize::value_type &i : pnode->mapRecvBytesPerMsgCmd)
             mapRecvBytesMsgStats[i.first] += i.second;
+        LOCK(pnode->cs_vSend);
         for (const mapMsgCmdSize::value_type &i : pnode->mapSendBytesPerMsgCmd)
             mapSentBytesMsgStats[i.first] += i.second;
```
Comment (UdjinM6) [inline src/net.cpp:1409]: 20816: could reduce the scope of locks here

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
20584:
1. missing changes in `src/validation.h`
2. missing changes in `src/logging.*` because we have no 18669 backport yet, so 20584 should be partial

14501: missing changes in `configure.ac`

File: src/net.cpp (hunk 1 above)
Comment (vijaydasmp) [reply src/net.cpp:1409]: Hello @UdjinM6 requesting suggestion for apt place for LOCK

Comment (UdjinM6) [reply src/net.cpp:1409]:
```diff
diff --git a/src/net.cpp b/src/net.cpp
index a7d3c4183f..c85ae2eda7 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -1401,12 +1401,16 @@ void CConnman::CalculateNumConnectionsChangedStats()
     mapSentBytesMsgStats[NET_MESSAGE_COMMAND_OTHER] = 0;
     auto vNodesCopy = CopyNodeVector(CConnman::FullyConnectedOnly);
     for (auto pnode : vNodesCopy) {
-        LOCK(pnode->cs_vRecv);
-        for (const mapMsgCmdSize::value_type &i : pnode->mapRecvBytesPerMsgCmd)
-            mapRecvBytesMsgStats[i.first] += i.second;
-        LOCK(pnode->cs_vSend);
-        for (const mapMsgCmdSize::value_type &i : pnode->mapSendBytesPerMsgCmd)
-            mapSentBytesMsgStats[i.first] += i.second;
+        {
+            LOCK(pnode->cs_vRecv);
+            for (const mapMsgCmdSize::value_type &i : pnode->mapRecvBytesPerMsgCmd)
+                mapRecvBytesMsgStats[i.first] += i.second;
+        }
+        {
+            LOCK(pnode->cs_vSend);
+            for (const mapMsgCmdSize::value_type &i : pnode->mapSendBytesPerMsgCmd)
+                mapSentBytesMsgStats[i.first] += i.second;
+        }
         if(pnode->fClient)
             spvNodes++;
         else
```

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Review by knst — CHANGES_REQUESTED** [review]
for backport #20584 there are several missing changes:
 - feebump. We won't have feebump -> ignore these changes
 - changes related to log-categories that has been refactored in  #18669 . Please, pick this backport, it's done here: https://github.com/dashpay/dash/pull/5541/commits/de17e4a46ebcd709da0a3732e1af92986712bb4e
 - `const NodeContext& node = EnsureNodeContext(context);` - we have different implementation of EnsureNodeContext<> and related methods. Changes are impossible to apply.

just update `EnsureAnyNodeContext` a bit like that, `const` is needed ONLY there:
```cpp
-    auto* node_context = GetContext<NodeContext>(context);
+    auto* const node_context = GetContext<NodeContext>(context);
```

Please, mark this backport as full; not partial.

**Comment by knst** [comment]
```
-std::vector<LogCategory> LogCategoriesList();
+std::vector<LogCategory> LogCategoriesList() const;
-std::string LogCategoriesString()
+std::string LogCategoriesString() const
```
from https://github.com/bitcoin/bitcoin/pull/20584 for https://github.com/bitcoin/bitcoin/pull/18669
@vijaydasmp

**Review by knst — APPROVED** [review]
utACK

**Comment by vijaydasmp** [comment]
Hello @knst , I had missed **const** after std::vector<LogCategory> BCLog::Logger::LogCategoriesList(), added that and pushed changes , please have a look

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**